src/DSA.jl

```diff
@@ -1,174 +1,313 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct DSADriverState <: AbstractMaterialState
-    time::Float64 = zero(Float64)
-    strain::SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module DSAModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericDSA, GenericDSADriverState, GenericDSAParameterState, GenericDSAVariableState
+
+# specialization for Float64
+export DSA, DSADriverState, DSAParameterState, DSAVariableState
+
+@with_kw mutable struct GenericDSADriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for DSA (dynamic strain aging) material.
+
+This is similar to the Chaboche model, but with additional static recovery terms.
+
+`E`: Young's modulus
+`nu`: Poisson's ratio
+`R0`: initial yield strength
+`Kn`: plasticity multiplier divisor (drag stress)
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q`: shift parameter for yield strength evolution
+`b`: multiplier for yield strength evolution
+`w`: ???
```

Comment:
A parameter that controls the average waiting time dislocation is arrested at localized obstacles.
Represents a strain increment produced when all arrested dislocations overcome localized obstacles and move towards to the next pinned configuration. In practice, this parameter controls how fast the effective aging time reacts to plastic strain.
\dot{t}_a = 1 - t_a \dot{p} / w